File: adt2sys/sdl/pas/draw_screen.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure draw_screen;

const
   frame_start: Longint = 0;
   frame_end: Longint = 0;
   actual_frame_end: Longint = 0;

begin
  realtime_gfx_poll_proc;
  draw_screen_proc;
  If _update_sdl_screen then SDL_Flip(screen);

  If _draw_screen_without_delay then _draw_screen_without_delay := FALSE
  else begin // keep framerate
         actual_frame_end := SDL_GetTicks;
         frame_end := frame_start+(1000 DIV sdl_frame_rate);
         // always sleep at least 2 msec
         If (actual_frame_end+2 > frame_end) then frame_end := actual_frame_end+2;
         SDL_Delay(frame_end-actual_frame_end);
         frame_start := SDL_GetTicks;
       end;
end;

File: adt2sys/sdl/draw_screen.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

static int32_t frame_start = 0;

void draw_screen (void)
{
  realtime_gfx_poll_proc ();
  draw_screen_proc ();
  if (_update_sdl_screen)
    SDL_Flip (screen);

  if (_draw_screen_without_delay)
    _draw_screen_without_delay = false;
  else
  {
    int32_t actual_frame_end;
    int32_t frame_end;

    // keep framerate
    actual_frame_end = SDL_GetTicks ();
    frame_end = frame_start + (1000 / sdl_frame_rate);
    // always sleep at least 2 msec
    if (frame_end < actual_frame_end + 2)
      frame_end = actual_frame_end + 2;
    SDL_Delay (frame_end - actual_frame_end);
    frame_start = SDL_GetTicks ();
  }
}
